### PR TITLE
HttpListener version for baseline comparisions

### PR DESCRIPTION
**Pull request by jeffschwMSFT**
We have been discussing what a node like version of this would look like (eg. framework less).  This is a proposed min version that could be used for comparisons.

### CODE REVIEW

File: AlbumsNetCoreNoFramework/Program.cs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+[JsonSourceGenerationOptions(WriteIndented = true)]
+[JsonSerializable(typeof(Album))]
+partial class AlbumJsonSerializerContext : JsonSerializerContext
+{
+}
+
+[JsonSourceGenerationOptions(WriteIndented = true)]
+[JsonSerializable(typeof(Album[]))]
+partial class AlbumArrayJsonSerializerContext : JsonSerializerContext
```
Comment (eerhardt) [inline AlbumsNetCoreNoFramework/Program.cs]: (Nit) can this be collapsed into the previous json context?

Comment (jeffschwMSFT) [reply AlbumsNetCoreNoFramework/Program.cs]: Good point, I was learning the source generator and did not revisit.  Fixed

File: AlbumsNetCoreNoFramework/AlbumsNetCoreNoFramework.csproj [hunk 2]
```diff
@@ -0,0 +1,10 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>net7.0</TargetFramework>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <Nullable>enable</Nullable>
```
Comment (eerhardt) [inline AlbumsNetCoreNoFramework/AlbumsNetCoreNoFramework.csproj:7]: Can you set the same properties in this csproj as the AlbumsNetCore app?

Comment (jeffschwMSFT) [reply AlbumsNetCoreNoFramework/AlbumsNetCoreNoFramework.csproj:7]: Done

File: AlbumsNetCoreNoFramework/Program.cs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+{
+    new Album("1", "Pawn Hearts", "Van der Graaf Generator", 26.99),
+    new Album("2", "A Passion Play", "Jethro Tull", 17.99),
+    new Album("3", "Tales from Topographic Oceans", "Yes", 32.99)
+};
+
+// listen
+while(http.IsListening)
+{
+    var context = await http.GetContextAsync();
```
Comment (davidfowl) [inline AlbumsNetCoreNoFramework/Program.cs]: A couple of things and it's why this model is so hard to use:
- This needs to be disposed to handle the next request.
- It doesn't handle concurrent requests. You need to dispatch to get concurrency here.

File: AlbumsNetCoreNoFramework/Program.cs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    {
+        // POST
+        if (context.Request.HttpMethod.Equals("POST",  StringComparison.OrdinalIgnoreCase))
+        {
+            // read content
+            Album album;
+            using (var reader = new StreamReader(context.Request.InputStream, context.Request.ContentEncoding))
+            {
+                album = (Album)JsonSerializer.Deserialize(reader.ReadToEnd(), typeof(Album), AlbumJsonSerializerContext.Default);
+            }
```
Comment (davidfowl) [inline AlbumsNetCoreNoFramework/Program.cs]: More efficient and async:

```suggestion
            var album = JsonSerializer.DeserializeAsync<Album>(context.Request.InputStream, AlbumJsonSerializerContext.Default);
```

File: AlbumsNetCoreNoFramework/Program.cs [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+[JsonSourceGenerationOptions(WriteIndented = true)]
+[JsonSerializable(typeof(Album))]
+[JsonSerializable(typeof(Album[]))]
+partial class AlbumJsonSerializerContext : JsonSerializerContext
+{
+}
+
+// Album represents data about a record album.
+internal record Album(string Id, string Title, string Artist, double Price);
```
Comment (davidfowl) [inline AlbumsNetCoreNoFramework/Program.cs:91]: Why is the id a string here?

File: AlbumsNetCoreNoFramework/Program.cs (hunk 4 above)
Comment (jeffschwMSFT) [reply AlbumsNetCoreNoFramework/Program.cs]: @eerhardt are the DesrializeAsync&lt;Type&gt; methods available for source generation?  DeserializeAsync(stream) is and I have changed to that, but I still need to unbox

Comment (jeffschwMSFT) [reply AlbumsNetCoreNoFramework/Program.cs]: scratch that, I think I see what I was doing wrong.

File: AlbumsNetCoreNoFramework/Program.cs (hunk 3 above)
Comment (jeffschwMSFT) [reply AlbumsNetCoreNoFramework/Program.cs]: +1

There are tradeoffs.  Along the lines of doing it manual, I added manual concurrency. :)

File: AlbumsNetCoreNoFramework/Program.cs [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    {
+        // POST
+        if (context.Request.HttpMethod.Equals("POST",  StringComparison.OrdinalIgnoreCase))
+        {
+            // read content
+            Album album;
+            using (var reader = new StreamReader(context.Request.InputStream, context.Request.ContentEncoding))
+            {
+                album = await JsonSerializer.DeserializeAsync<Album>(context.Request.InputStream, AlbumJsonSerializerContext.Default.Album);
+            }
```
Comment (davidfowl) [inline AlbumsNetCoreNoFramework/Program.cs]:
```suggestion
            Album album = await JsonSerializer.DeserializeAsync<Album>(context.Request.InputStream, AlbumJsonSerializerContext.Default.Album);
```

File: AlbumsNetCoreNoFramework/Program.cs [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    context.Response.ContentLength64 = buffer.Length;
+    context.Response.ContentType = contentType;
+    using (var output = context.Response.OutputStream)
+    {
+        await output.WriteAsync(buffer, 0, buffer.Length);
+    }
+}
+
+// GetAlbums responds with the list of all albums as JSON.
+string GetAlbums() => JsonSerializer.Serialize(myAlbums, typeof(Album[]), AlbumJsonSerializerContext.Default);
```
Comment (davidfowl) [inline AlbumsNetCoreNoFramework/Program.cs]: Do you need typeof(Album[]) there? The generic overload should work right?

File: AlbumsNetCoreNoFramework/Program.cs [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+#pragma warning disable 4014 // no await
+for(var i=0; i<concurrency; i++) http.GetContextAsync().ContinueWith(ProcessRequestHandler);
+#pragma warning disable 4014
+
+System.Console.ReadLine();
+
+async void ProcessRequestHandler(Task<HttpListenerContext> result)
+{
+    var context = await result;
```
Comment (davidfowl) [inline AlbumsNetCoreNoFramework/Program.cs:31]: This needs to be disposed.

```suggestion
    using var context = await result;
```

File: AlbumsNetCoreNoFramework/Program.cs [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+            if (idindx > 0 && idindx != context.Request.RawUrl.Length - 1) responseString = GetAlbumById(context.Request.RawUrl.Substring(idindx+1));
+            else responseString = GetAlbums();
+        }
+    }
+    else responseString = "not found";
+
+    // write details
+    var buffer = System.Text.Encoding.UTF8.GetBytes(responseString);
+    context.Response.ContentLength64 = buffer.Length;
+    context.Response.ContentType = contentType;
```
Comment (davidfowl) [inline AlbumsNetCoreNoFramework/Program.cs]: Few more problems:

- This isn't writing JSON to the response; it's writing a string when there's a 404.
- The status code should be 404 (in the not found case)

**Comment by eerhardt** [comment]
@jeffschwMSFT - I can't get the app to work in a docker container. From the `AlbumsNetCoreNoFramework` folder, I ran:

`docker build . -t albums; docker run -it --rm -p 5000:5000 --name albums albums`

The container built and started, but it actively refused connection:

image

**Comment by jeffschwMSFT** [comment]
I anticipated that.  The HttpListener is local host.  I can push a change to make it work for dns.

**Comment by davidfowl** [comment]
Make the host *

**Comment by eerhardt** [comment]
> Make the host *

That works inside the Docker container, but it doesn't work when `dotnet run`ing on Windows:

```
Unhandled exception. System.Net.HttpListenerException (5): Access is denied.
   at System.Net.HttpListener.AddPrefixCore(String registeredPrefix)
   at System.Net.HttpListener.AddAllPrefixes()
   at System.Net.HttpListener.Start()
   at Program.<Main>$(String[] args) in C:\git\Albums\AlbumsNetCoreNoFramework\Program.cs:line 14  
```

**Comment by jeffschwMSFT** [comment]
This is the same when you have the ip address.  HttpListener need permission to listen on the external address

**Comment by jeffschwMSFT** [comment]
How does kestrel listen without elevation?

**Comment by eerhardt** [comment]
```C#
var host = OperatingSystem.IsWindows() ? "127.0.0.1" : "*";
```

Seems to work for me.  Objection on me pushing that change?

**Comment by jeffschwMSFT** [comment]
Interesting, windows requires elevation.

fwiw, here is the way to avoid the wild card (which is stated as not a best practice, but good for experimenting).

```
string GetHost(bool local)
{
    var host = local ? Dns.GetHostEntry("localhost") : Dns.GetHostEntry(Dns.GetHostName());
    IPAddress ip = null;

    // look for appropriate ipv4 address
    for(var i=0; i<host.AddressList.Length && ip is null; i++)
    {
        if (host.AddressList[i].AddressFamily == AddressFamily.InterNetwork) 
            ip = host.AddressList[i];
    }

    if (ip is null && host.AddressList.Length > 0) ip = host.AddressList[0];
    if (ip is null) return "127.0.0.1";

    return ip.ToString();
}
```

**Comment by eerhardt** [comment]
> here is the way to avoid the wild card (which is stated as not a best practice, but good for experimenting).

That code also doesn't seem to work inside a docker container.

**Comment by jeffschwMSFT** [comment]
Interesting, I wonder what the hostname returns.  I will look deeper.

**Comment by jeffschwMSFT** [comment]
I ran a quick experiment and the Dns type is not returning an external host ip from within the container.  It only has a loop back address.  I think your proposed change for "*" is the best option.

**Comment by eerhardt** [comment]
OK, I pushed that change, and added the HttpListener to the crank benchmarks. Here are the results of the HttpListener version compared to the NativeAOT Minimal APIs app (with and without Server GC):

| application         | httplistener-alpine | net-alpine-serverGC |          | net-alpine-workstationGC |          |
| ------------------- | ------------------- | ------------------- | -------- | ------------------------ | -------- |
| CPU Usage (%)       |                  40 |                  93 | +132.50% |                       66 |  +65.00% |
| Cores usage (%)     |                 483 |               1,111 | +130.02% |                      792 |  +63.98% |
| Working Set (MB)    |                  38 |                 240 | +531.58% |                       29 |  -23.68% |
| Build Time (ms)     |              14,485 |              30,851 | +112.99% |                   30,528 | +110.76% |
| Start Time (ms)     |                  79 |                  87 |  +10.13% |                       92 |  +16.46% |
| Published Size (KB) |              23,503 |              41,762 |  +77.69% |                   41,442 |  +76.33% |
| Working Set (75th)  |                  37 |                 130 | +251.35% |                       29 |  -21.62% |


| load                   | httplistener-alpine | net-alpine-serverGC |          | net-alpine-workstationGC |          |
| ---------------------- | ------------------- | ------------------- | -------- | ------------------------ | -------- |
| CPU Usage (%)          |                  42 |                  88 | +109.52% |                       71 |  +69.05% |
| Cores usage (%)        |                 501 |               1,052 | +109.98% |                      855 |  +70.66% |
| Start Time (ms)        |                  79 |                  79 |    0.00% |                       80 |   +1.27% |
| First Request (ms)     |                  57 |                  64 |  +12.28% |                       64 |  +12.28% |
| Requests               |           1,478,010 |           5,018,905 | +239.57% |                3,938,188 | +166.45% |
| Bad responses          |                   0 |                   0 |          |                        0 |          |
| Latency 50th (ms)      |                1.97 |                0.65 |  -66.84% |                     0.70 |  -64.45% |
| Latency 75th (ms)      |                3.64 |                0.79 |  -78.46% |                     0.96 |  -73.63% |
| Latency 90th (ms)      |                5.18 |                0.96 |  -81.36% |                     1.72 |  -66.78% |
| Latency 95th (ms)      |                6.51 |                1.15 |  -82.37% |                     3.27 |  -49.76% |
| Latency 99th (ms)      |                9.40 |                2.89 |  -69.31% |                     4.25 |  -54.85% |
| Mean latency (ms)      |                2.59 |                0.76 |  -70.62% |                     0.97 |  -62.52% |
| Max latency (ms)       |               59.62 |               38.28 |  -35.80% |                    67.53 |  +13.26% |
| Requests/sec           |              98,734 |             335,513 | +239.81% |                  262,908 | +166.28% |
| Requests/sec (max)     |             147,584 |             387,901 | +162.83% |                  318,477 | +115.79% |
| Read throughput (MB/s) |               22.29 |               76.58 | +243.62% |                    60.08 | +169.58% |

**Comment by jeffschwMSFT** [comment]
Thanks Eric.  The most interesting line is the following:

<html>
<body>


Published Size (KB) | 23,503 | 41,762 | +77.69% | 41,442 | +76.33%
-- | -- | -- | -- | -- | --



</body>
</html>

Great to see that working set is not really a win with the noframework version.

I think that beyond native aot, that publish flags like '/p:PublishSingleFile=true /p:PublishTrimmed=true /p:EnableCompressionInSingleFile=true' would be the most competitive on size-on-disk.

**Pull request merged**